src/rnpkeys/tui.c

```diff
@@ -157,9 +161,40 @@ rnp_generate_key_expert_mode(rnp_t *rnp)
         rnp->action.generate_key_ctx.rsa.modulus_bit_len = ask_bitlen(input_fd);
         break;
     case PGP_PKA_ECDH:
-    case PGP_PKA_ECDSA:
+    case PGP_PKA_ECDSA: {
         rnp->action.generate_key_ctx.ecc.curve = ask_curve(input_fd);
-        break;
+        if (PGP_HASH_UNKNOWN == key_desc->hash_alg) {
+            return PGP_E_ALG_UNSUPPORTED_HASH_ALG;
+        }
+
+        size_t digest_length = 0;
+        if (!pgp_digest_length(key_desc->hash_alg, &digest_length)) {
+            // Implementation error
+            return PGP_E_FAIL;
+        }
+
+        // Adjust hash to curve
+        switch (key_desc->ecc.curve) {
+        case PGP_CURVE_NIST_P_256:
+            if (digest_length < 32) {
+                key_desc->hash_alg = PGP_HASH_SHA256;
+            }
+            break;
+        case PGP_CURVE_NIST_P_384:
+            if (digest_length < 48) {
```

Comment:
It kind of comes from the fact how ecdsa works. But I agree, it's maybe not obvious and even if it is, additional comment doesn't hurt.
Done